CONTRIBUTING.md

```diff
@@ -0,0 +1,183 @@
+  character.
+- Use tabs for indentation and spaces for alignment for all C source code and
+  header files.
+- Trim all trailing whitespace.
+- Where multiple words must be combined into one, use hyphens in strings,
+  filenames, and documentation; however, use underscores in function, variable
+  names and header guards.
+
+### C files
+
+Functions in drivers should be topologically sorted (i.e. the `probe` function
+belongs at the bottom). Other functions should be sorted in alphabetical order
+with small, `static` helper functions at the top. Functions and macros in
+headers should be alphabetically sorted whenever possible.
+
+### Writing headers
+
+- Header files act as APIs and are reserved for functions that will be
+  _exposed_ to the rest of the project. Macros, `struct` definitions, and
+  function declarations that are only utilized by a driver or module belong in
+  the corresponding implementation file, rather than the header file.
+- Public macros, function prototypes, and declarations should be documented
+  according to [Doxygen JavaDoc-style comment blocks][doxygen].
+- Functions defined in assembly should also be declared in a header, unless
+  they are not intended to be called directly (e.g. `__udivsi3`).
+- Ideally, the header should be named the same as the driver or source file,
+  except that it should be placed inside the include hierarchy.
+- The header guard should be the same as the path you would need to put in your
+  `#include` directive, but with the following changes:
+  - All letters should be uppercase.
+  - All hyphens, slashes, and periods should be replaced by underscores.
+
+### Using Headers
+
+- Drivers are not allowed to include platform headers.
+- Library code is only allowed to include library headers.
+
+### Naming in drivers
+
+Drivers should be named as vaguely as the hardware allows. For example, if a
+driver supports the CCU on all sunxi devices, it should be named `sunxi-ccu`.
+If it supports the 64-bit counter _first introduced_ with the sun6i series, it
+should be named `sun6i-cnt64`. If a driver only supports the SY8106A voltage
+regulator, it should be named `sy8106a`.
+
+### Assembly
+
+Assembly files are preprocessed, so they should use the file extension `.S`
+(and this is the only extension the build system looks for). They should not
+have the same path as a C file (excluding the extension), as the build system
+cannot disambiguate the two.
+
+Assembly should be formatted as follows:
+- Local control-flow labels should be numbered, not named.
+- Local data labels should be named descriptively.
+- Global data and functions should use the `data` and `func`/`endfunc` macros
+  from `macros.S`.
+- Indent the instruction by one tab.
+- Indent the first operand with one tab.
+- Indent following operands with a single space.
+- Use `#` to start comments.
+- Align comments at 40 columns with tabs.
+
+### Continuous Integration Testing
```

Comment:
Maybe this should be moved under the "Submitting a Patch" heading?

Reply:
Done.